### PR TITLE
Add support for use_tools on chat block in Dust App

### PR DESCRIPTION
**Pull request by flvndvd**
## Description

This PR establishes the groundwork for our upcoming refactoring of the OpenAI model integration. We're transitioning from using functions to tools. To mitigate potential risks associated with this release, we'll maintain two separate code paths, which will be controlled by a `use_tools` setting available in both the API and the user interface.

image

### CODE REVIEW

File: core/src/blocks/chat.rs [hunk 1]
```diff
@@ -292,6 +292,11 @@ impl Block for Chat {
                     extras["response_format"] = json!(s.clone());
                 }
 
+                extras["use_tools"] = json!(v.get("use_tools").map_or(false, |v| match v {
```
Comment (fontanierh) [inline core/src/blocks/chat.rs]: `map_or` does not seem useful here, can we simply match against `json!(...)` ?

File: front/components/app/blocks/Chat.tsx [hunk 2]
```diff
@@ -460,22 +466,41 @@ export default function Chat({
...
-                      readOnly
-                        ? "border-white ring-0 focus:border-white focus:ring-0"
-                        : "border-white focus:border-gray-300 focus:ring-0"
-                    )}
-                    spellCheck={false}
-                    readOnly={readOnly}
-                    value={block.config.function_call}
-                    onChange={(e) => handleFunctionCallChange(e.target.value)}
-                  />
+              <div className="flex flex-row">
```
Comment (fontanierh) [inline front/components/app/blocks/Chat.tsx:469]: do we want to get this to us only ?

**Review by fontanierh — APPROVED** [review]
looks good modulo visibility of the new feature

File: core/src/blocks/chat.rs (hunk 1 above)
Comment (flvndvd) [reply core/src/blocks/chat.rs]: Fixed.

**Comment by spolu** [comment]
Why the use_tools option? We should have only one way to use the chat block that maps to different API for different models.

As a user it’s also super hard to understand what it means.

The end goal is to have only one path right?

The problem of adding such option is that people will click on them and migrating specs is super hard.

I would suggest introducing a secret syntax for testing in the return of the functions specification instead if you want to have a way to test internally 

Please do not merge as is 🙏

**Comment by spolu** [comment]
I see you go through the config which is better than spec but still our users should not be faced with this option. Our largest enterprise users use Dust apps all day; this should not be considered as a product surface with lower quality (all the contrary)

**Comment by fontanierh** [comment]
@spolu We need a proper way to test with real world assistant and doing so locally with our remove dust apps setup is very hard

**Comment by flvndvd** [comment]
> The end goal is to have only one path right?

100 percents!

The goal as explained in the PR is to de-risk the release of tools. This code will be removed once it's done.

**Comment by flvndvd** [comment]
@spolu good for you?

**Comment by spolu** [comment]
Still let’s not expose a button to all users

**Comment by spolu** [comment]
Testing can be done in dev + the underlying mechanism is the same for OpenAI so there will be no difference of behavior.

Why not just move to it?

**Comment by fontanierh** [comment]
> Testing can be done in dev + the underlying mechanism is the same for OpenAI so there will be no difference of behavior.

> Why not just move to it?

Testing end to end fully locally with assistants that do retrieval is very complicated

**Comment by fontanierh** [comment]
> underlying mechanism is the same for OpenAI so there will be no difference of behavior.

We are not sure that it won't break or alter performance due to the fact that we don't show a `function_call` for every `function` we render. Given that this is maximal blast radius it is nice to do a real test.

**Comment by spolu** [comment]
I would argue that we don’t need to test this do we?

**Review by spolu — COMMENTED** [review]
👍

**Pull request merged**